### DISCUSSION TITLE
Error when commiting changes to converter file, exit code 1.

### DISCUSSION THREAD

**Discussion by ok-67** [Koenkk/zigbee2mqtt#21715] [Devices]
I have made some changes to an converter (for a Namron thermostat outlet) which runs OK both at my computer with updated sources locally and as an external converter under Home Assistant. However when I try to commit my changes to the sources I get the following error, and I cant figure why (pars of the code is shown further down). 

What am I doing wrong?

_
Run npm test
  npm test
  shell: /usr/bin/bash -e {0}
  env:
    NPM_CONFIG_USERCONFIG: /home/runner/work/_temp/.npmrc
    NODE_AUTH_TOKEN: XXXXX-XXXXX-XXXXX-XXXXX

> zigbee-herdsman-converters@18.46.0 test
> ZHC_TEST=true jest test

PASS test/modernExtend.test.ts (30.583 s)
PASS test/generateDefinition.test.ts (30.664 s)
PASS test/lumi.test.js
PASS test/color.test.js
PASS test/philips.test.js
PASS test/sonoff.test.ts
PASS test/utils.test.js
PASS test/fromZigbee.test.js
PASS test/otaCommon.test.js
FAIL test/index.test.js (34.815 s)
  ● index.js › Exposes access matches toZigbee

    4512749-N - local_temperature, supports set: true

      449 |
      450 |                     if ((expose.access & exposes.access.SET) != (toZigbee && toZigbee.convertSet ? exposes.access.SET : 0)) {
    > 451 |                         throw new Error(`${device.model} - ${property}, supports set: ${!!(toZigbee && toZigbee.convertSet)}`);
          |                               ^
      452 |                     }
      453 |
      454 |                     if ((expose.access & exposes.access.GET) != (toZigbee && toZigbee.convertGet ? exposes.access.GET : 0)) {

      at test/index.test.js:451:31
          at Array.forEach (<anonymous>)
      at Object.<anonymous> (test/index.test.js:427:27)

Test Suites: 1 failed, 9 passed, 10 total
Tests:       1 failed, 31 skipped, 222 passed, 254 total
Snapshots:   0 total
Time:        35.137 s
Ran all test suites matching /test/i.
Error: Process completed with exit code 1._

Part of the code:
_
   {
        zigbeeModel: ['4512749-N'],
        model: '4512749-N',
        vendor: 'Namron',
        description: 'Namron Thermostat outlet socket',
        fromZigbee: [fz.thermostat],
        toZigbee: [tzLocal.namronPlug],
        extend: [
            onOff(),
            temperature(),
            electricityMeter(),
            binary({
                name: 'operation_mode',
                cluster: 0x0000,
                attribute: {ID: 0x1000, type: 0x30},
                description: 'Set operation mode as ON/OFF or as Thermostat.',
                valueOn: ['Thermostat', 0x01],
                valueOff: ['ON/OFF', 0x00],
                zigbeeCommandOptions: {manufacturerCode: 0x1224},
            }),
        ],
        exposes: [
            e.climate()
                .withSetpoint('occupied_heating_setpoint', 5, 35, 0.5)
                .withLocalTemperature()
                .withLocalTemperatureCalibration(-2.5, 2.5, 0.1)
                .withSystemMode(['off', 'heat'])
                .withRunningState(['idle', 'heat']),
        ],
        configure: async (device, coordinatorEndpoint, logger) => {
            const endpoint2 = device.getEndpoint(2);
            await reporting.bind(endpoint2, coordinatorEndpoint, ['hvacThermostat']);

            // Thermostat reporting
            await reporting.thermostatOccupiedHeatingSetpoint(endpoint2);
            await reporting.thermostatTemperature(endpoint2, {min: 1, change: 10});
            await reporting.thermostatSystemMode(endpoint2);
            await reporting.thermostatTemperatureCalibration(endpoint2);
            await reporting.thermostatRunningState(endpoint2);
            await endpoint2.read('hvacThermostat', ['systemMode', 'runningState', 'occupiedHeatingSetpoint']);
        },
    ota: ota.zigbeeOTA,
    },
_

**Comment by Koenkk** [comment]
Can you submit the PR, then I will look at the test

**Comment by ok-67** [comment]
[namron.ts.zip](https://github.com/Koenkk/zigbee2mqtt/files/14549511/namron.ts.zip)
This is the modified namron.ts file that I like to add/modify

**Comment by ok-67** [comment]
I'm have an idea about what is the problem. If I remove this:

        exposes: [
            e.climate()
                .withSetpoint('occupied_heating_setpoint', 5, 35, 0.5)
                .withLocalTemperature()
                .withLocalTemperatureCalibration(-2.5, 2.5, 0.1)
                .withSystemMode(['off', 'heat'])
                .withRunningState(['idle', 'heat']),
        ],

from the definition it seems to pass the test, but then I can't set the thermostat temperature etc.
But at least I have figured out what is making the test not to pass.

**Comment by ok-67** [comment]
If I remove running_state and local_temperature from the definition, it seems to pass the test, at least locally on my computer, but I have no clue about why and what actually is the problem.

    {
        zigbeeModel: ['4512749-N'],
        model: '4512749-N',
        vendor: 'Namron',
        description: 'Namron Thermostat outlet socket',
        fromZigbee: [fz.thermostat],
        toZigbee: [
            tzLocal.namronPlug,
            tz.thermostat_occupied_heating_setpoint,
            // tz.thermostat_local_temperature,
            tz.thermostat_local_temperature_calibration,
            tz.thermostat_system_mode,
            // tz.thermostat_running_state,
        ],
        extend: [
            onOff(),
            temperature(),
            electricityMeter(),
            binary({
                name: 'operation_mode',
                cluster: 0x0000,
                attribute: {ID: 0x1000, type: 0x30},
                description: 'Set operation mode as ON/OFF or as Thermostat.',
                valueOn: ['Thermostat', 0x01],
                valueOff: ['ON/OFF', 0x00],
                zigbeeCommandOptions: {manufacturerCode: 0x1224},
            }),
        ],
        exposes: [
            e.climate()
                .withSetpoint('occupied_heating_setpoint', 5, 35, 0.5)
                // .withLocalTemperature()
                .withLocalTemperatureCalibration(-2.5, 2.5, 0.1)
                .withSystemMode(['off', 'heat']),
            // .withRunningState(['idle', 'heat']),
        ],
        configure: async (device, coordinatorEndpoint, logger) => {
            const endpoint2 = device.getEndpoint(2);
            await reporting.bind(endpoint2, coordinatorEndpoint, ['hvacThermostat']);

            // Thermostat reporting
            await reporting.thermostatOccupiedHeatingSetpoint(endpoint2);
            await reporting.thermostatTemperature(endpoint2, {min: 1, change: 10});
            await reporting.thermostatSystemMode(endpoint2);
            await reporting.thermostatTemperatureCalibration(endpoint2);
            await reporting.thermostatRunningState(endpoint2);
            await endpoint2.read('hvacThermostat', [
                'occupiedHeatingSetpoint',
                // 'local_temperature',
                'localTemperatureCalibration',
                'systemMode',
                // 'runningState',
            ]);
            // localTemperature
        },
        ota: ota.zigbeeOTA,
    },

**Comment by ok-67** [comment]
Changing the toZigbe statement from

        toZigbee: [
            tzLocal.namronPlug,
            tz.thermostat_occupied_heating_setpoint,
            tz.thermostat_local_temperature,
            tz.thermostat_local_temperature_calibration,
            tz.thermostat_system_mode,
            tz.thermostat_running_state,
        ],

to

        toZigbee: [
            tz.thermostat_occupied_heating_setpoint,
            tz.thermostat_local_temperature,
            tz.thermostat_local_temperature_calibration,
            tz.thermostat_system_mode,
            tz.thermostat_running_state,
            tzLocal.namronPlug,
        ],

seems to solve the problem.

Can someone explain why??? (I am new to typescript/javascript so an explanation would be nice).

I will run some more tests with my code, since I did seen an error showing up after some time,  before I finally submit the code.

**Comment by ok-67** [comment]
Unfortunately, this change to the code, actually breaks the code. :-( 
It seems like it is calling the standard thermostat routines instead of the ones that I have made, which reads from endpoint 2, where the thermostat cluster of this socket is located. So either the code works and don't pass the test or it passes the test and don't work...

**Comment by ok-67** [comment]
Silly me :-( 
I have not implemented local_temperature and running_state. But then again, you can not set the local temperature... Hmmmmm
